lib/changesreader.js

```diff
@@ -0,0 +1,284 @@
+const EventEmitter = require('events').EventEmitter
+const stream = require('stream')
+const EVENT_BATCH = 'batch'
+const EVENT_CHANGE = 'change'
+const EVENT_SEQ = 'seq'
+const EVENT_ERROR = 'error'
+
+// streaming line breaker
+const liner = () => {
+  const liner = new stream.Transform({ objectMode: true })
+
+  liner._transform = function (chunk, encoding, done) {
+    let data = chunk.toString('utf8')
+    if (this._lastLineData) {
+      data = this._lastLineData + data
+      this._lastLineData = null
+    }
+
+    const lines = data.split(/\s*\n/)
+    this._lastLineData = lines.splice(lines.length - 1, 1)[0]
+    lines.forEach(this.push.bind(this))
+    done()
+  }
+
+  liner._flush = function (done) {
+    this.push(this._lastLineData)
+    this._lastLineData = null
+    done()
+  }
+
+  return liner
+}
+
+// streaming change processor
+const changeProcessor = (ee, batchSize) => {
+  const changeProcessor = new stream.Transform({ objectMode: true })
+  const buffer = []
+  changeProcessor.lastSeq = '0'
+
+  changeProcessor._transform = function (chunk, encoding, done) {
+    // remove last char from string
+    if (chunk[chunk.length - 1] === ',') {
+      chunk = chunk.slice(0, -1)
+    }
+
+    try {
+      const j = JSON.parse(chunk)
+      buffer.push(j)
+      if (buffer.length >= batchSize) {
+        ee.emit(EVENT_BATCH, buffer.splice(0, batchSize))
+      }
+      done()
+    } catch (e) {
+      // look for last_seq
+      const match = chunk.match(/"last_seq":"([^"]+)"/)
```

Comment:
What happens if the server replies with a number? Like in ..

```json
{
  "last_seq": 1234
}
```

Reply:
Oh couchdb says that `last_seq` type is json!

> **last_seq** (json) – Last change update sequence

Reply:
yeah string should be safe here.

Reply:
I believe that the regex should be fixed, as it fail for servers responding with numbers.

For reference here's an example of a server that uses numbers: https://replicate.npmjs.com/

Reply:
👍 done

Reply:
Thanks!